Create User-Service

rbrmnv opened the pull request:
- Add basic User CRUD
- Configured migrations using Liquibase
- Configured Swagger
- Divided into modules: api, db, impl
- Configured SLF4J logging
- Covered with tests

rbrmnv committed:
Init commit for user service

rbrmnv committed:
Add Liquibase work via console

rbrmnv committed:
Add AspectJ logging to user-api module

Commit:
Apply fix from PR #2

Commit:
Fix :services:user-service Gradle configs

- I deleted all the settings from services/user-service/build.gradle because it's just a
  container module

- Add build.gradle to :services:user-services:user-impl

- Deleted duplicate settins from services/user-service/user-api/build.gradle and
  services/user-service/user-db/build.gradle

rbrmnv committed:
remove aop logic from services

rbrmnv committed:
move impl logic to user-impl module and switch doc language to english

rbrmnv committed:
add logging to controller and service

rbrmnv committed:
add test coverage

rbrmnv committed:
Merge branch 'develop' into feature/BRAEN-06/user-service-dev

E-Line21 commented on services/user-service/user-impl/src/main/java/com/technokratos/eateasy/userimpl/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+        return userRepository.findById(id)
+                .map(userMapper::toDto)
+                .orElseThrow(() -> new UserNotFoundException("User not found with id: " + id));
+    }
```
Использование конкатенации строк - крайне плохая практика. Лучше использовать `String.format("Hello, %s!", name)` или `"Hello, %s!".format(name)`

Можешь почитать об этом в [статье](https://www.yegor256.com/2014/06/19/avoid-string-concatenation.html)

rbrmnv replied:
изменил

E-Line21 commented on services/user-service/user-api/src/main/java/com/technokratos/eateasy/userapi/api/UserApi.java:
```diff
@@ -0,0 +1,85 @@
+    @Operation(summary = "Get all users", description = "Returns a list of all registered users")
+    @ApiResponse(responseCode = "200", description = "Successful request", content = @Content(schema = @Schema(implementation = UserResponseDto.class)))
+    @GetMapping
+    ResponseEntity<List<UserResponseDto>> getAllUsers();
```
Использование `ResponseEntity` избыточно и просто создает лишнюю обёртку. Лучше возвращать напрямую DTO с аннотациями `@ResponseBody` + `@ResponseStatus`

E-Line21 commented on services/user-service/user-impl/src/main/java/com/technokratos/eateasy/userimpl/controller/UserController.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.UUID;
+
+@RestController
+@RequestMapping("/users")
```
`@RequestMapping` должен быть над интерфейсом

E-Line21 commented on services/user-service/user-db/build.gradle:
```diff
@@ -0,0 +1,32 @@
+            classpath "$projectDir/db/changelog"
+            changelogFile 'db.changelog-master.yaml'
+            url 'jdbc:postgresql://localhost:5432/user_service_dev'
+            username 'postgres'
```
Параметры базы данных должны передаваться извне (переменные среды или properties файл)

E-Line21 commented on services/user-service/user-impl/build.gradle:
```diff
@@ -0,0 +1,24 @@
+    testImplementation 'org.mockito:mockito-junit-jupiter'
+    testImplementation("org.testcontainers:junit-jupiter")
+    implementation 'com.h2database:h2'
+    testImplementation("org.springframework.boot:spring-boot-starter-test")
```
- H2 подключен как `implementation`, хотя используется только в тестах. Должен быть `testImplementation`.
- `org.springframework.boot:spring-boot-starter-test` дублируется в зависимостях.

rbrmnv replied:
убрал вообще H2, буду контейнеры поднимать

E-Line21 commented on services/user-service/user-impl/src/main/java/com/technokratos/eateasy/userimpl/UserImplApplication.java:
```diff
@@ -0,0 +1,21 @@
+
+@SpringBootApplication
+@EntityScan("com.technokratos.eateasy.userimpl.model")
+@ComponentScan("com.technokratos.eateasy")
```
Лучше сканировать только пакеты своего модуля, чтобы не зацепить что-нибудь чужое

E-Line21 commented on services/user-service/user-impl/src/main/java/com/technokratos/eateasy/userimpl/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,25 @@
+    @ExceptionHandler(UserAlreadyExistsException.class)
+    public ResponseEntity<String> handleUserAlreadyExistsException(UserAlreadyExistsException ex) {
+        return ResponseEntity.status(HttpStatus.CONFLICT).body(ex.getMessage());
+    }
```
Обработку исключений `UserNotFoundException` и `UserAlreadyExistsException` можно унифицировать через общего родителя с полем `statusCode`

rbrmnv replied:
решил не трогать, тк делается глобальный обработчик

E-Line21 commented on services/user-service/user-impl/src/main/java/com/technokratos/eateasy/userimpl/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,25 @@
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<String> handleGeneralException(Exception ex) {
+        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(ex.getMessage());
+    }
```
Клиенту нельзя возвращать сырое сообщение об ошибке (`ex.getMessage()`), это может привести к утечке внутренней информации

rbrmnv replied:
также пропустим

E-Line21 replied:
Вообще я реализовал только обработку ошибок бизнес логики (`ServiceException`), всякие ошибки валидации и прочее я не обрабатываю

E-Line21 commented on services/user-service/user-impl/src/main/java/com/technokratos/eateasy/userimpl/controller/UserController.java:
```diff
@@ -0,0 +1,61 @@
+@Slf4j
+public class UserController implements UserApi {
+
+    private final UserService userService;
```
Достаточно просто `service`. Т.к. мы в `UserController`, то уже понятно, что это за сервис

E-Line21 commented on services/user-service/user-impl/src/main/java/com/technokratos/eateasy/userimpl/exception/GlobalExceptionHandler.java:
Нет обработки ошибок валидации (например, `MethodArgumentNotValidException`). В идеале вообще создать какой-нибудь класс `ErrorResponse`, чтобы при любых исключениях отвечать единым образом и не мучить фронтов

rbrmnv replied:
тоже пропустил

E-Line21 commented on services/user-service/user-api/src/main/java/com/technokratos/eateasy/userapi/api/UserApi.java:
```diff
@@ -0,0 +1,85 @@
+            @ApiResponse(responseCode = "404", description = "User not found")
+    })
+    @GetMapping(params = "email")
+    ResponseEntity<UserResponseDto> getUserByEmail(
```
Эта ручка используется Auth Service, когда пользователь входит в систему, и он ожидает, что в ответе будет также пароль пользователя для проверки

rbrmnv replied:
То есть мне нужно в респонзе возвращать пароль? Разве это безопасно?

E-Line21 replied:
Ну не пароль, а хэш. Почему не безопасно, это же внутренняя ручка? И как тогда ты предлагаешь его проверять? Я буду тебе и почту, и пароль передавать?

E-Line21 commented on services/user-service/user-api/src/main/java/com/technokratos/eateasy/userapi/dto/UserRequestDto.java:
```diff
@@ -0,0 +1,67 @@
+            requiredMode = Schema.RequiredMode.REQUIRED
+    )
+    @NotBlank(message = "First name cannot be blank")
+    private String firstName;
```
Лучше проверять длину текстового поля, чтобы туда "случайно" не загрузили "Войну и мир"

E-Line21 replied:
Для пароля лучше тоже ограничение в длину поставить

E-Line21 commented on services/user-service/user-impl/src/main/java/com/technokratos/eateasy/userimpl/model/UserEntity.java:
А как хранится роль пользователя?

rbrmnv replied:
Исправил всё

E-Line21 commented on services/user-service/user-impl/src/main/java/com/technokratos/eateasy/userimpl/service/UserService.java:
В названиях методов не нужно использовать "user". И так понятно, что UserService работает с пользователями

E-Line21 commented on services/user-service/user-impl/src/main/java/com/technokratos/eateasy/userimpl/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+    @Override
+    public UserResponseDto createUser(UserRequestDto userDto) {
+        if (userRepository.existsByUsername(userDto.getUsername())) {
+            log.warn("Username already exists: {}", userDto.getUsername());
```
Вряд ли это уровень логирования `warn`. Данная ошибка является стандартной для системы. Скорее, это `info` или даже `debug`.

rbrmnv replied:
Изменил все на debug

E-Line21 commented on services/user-service/user-impl/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,34 @@
+spring:
+  datasource:
+    url: jdbc:postgresql://localhost:5432/user_service_dev
+    username: postgres
```
Параметры базы данных лучше передавать извне

E-Line21 commented on services/user-service/user-api/.gitattributes:
Лучше удалить, этот файл не нужен

E-Line21 commented on services/user-service/user-api/.gitignore:
Лучше удалить, этот файл не нужен

E-Line21 commented on services/user-service/user-impl/src/main/java/com/technokratos/eateasy/userimpl/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+    @Override
+    public void deleteUser(UUID id) {
+        if (!userRepository.existsById(id)) {
+            log.warn("User not found with id: " + id);
```
Возможно, имеет смысл просто игнорировать этот случай, т.к. метод DELETE должен быть идемпотентным

E-Line21 commented on services/user-service/user-impl/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,34 @@
+    driver-class-name: org.postgresql.Driver
+  jpa:
+    hibernate:
+      ddl-auto: none
```
Вроде, здесь должно быть `validate`

E-Line21 commented on services/user-service/user-db/.gitattributes:
Лучше удалить, этот файл не нужен

E-Line21 commented on services/user-service/user-db/.gitignore:
Лучше удалить, этот файл не нужен

E-Line21 commented on services/user-service/user-api/build.gradle:
```diff
@@ -0,0 +1,13 @@
+}
+
+dependencies {
+	implementation 'org.springdoc:springdoc-openapi-starter-webmvc-ui:2.8.5'
```
Подключение `springdoc-openapi-starter-webmvc-ui` избыточно. Если другой сервис (который просто использует API-модуль для клиентских запросов) добавит его, как зависимость, у него неожиданно появится Swagger UI, хотя он, возможно, этого не хотел

E-Line21 commented on services/user-service/user-impl/src/main/java/com/technokratos/eateasy/userimpl/mapper/UserMapper.java:
```diff
@@ -0,0 +1,59 @@
+                        .build();
+    }
+
+    public void updateEntity(UserEntity entity, UserRequestDto dto) {
```
Метод `updateEntity` действительно SRP - маппер не должен заниматься обновлением сущностей, это ответственность сервисного слоя

rbrmnv replied:
Вот здесь не понимаю в чем вопрос

E-Line21 replied:
Как будто маппер не должен заниматься обновлением сущности, это нарушает Single Responsibility

rbrmnv committed:
delete unnecessary .git files
move sensitive data into environment variables
add role processing
rework UserApi
rework method names in service

E-Line21 commented on services/user-service/user-api/build.gradle:
```diff
@@ -6,7 +6,9 @@ java {
 
 dependencies {
-	implementation 'org.springdoc:springdoc-openapi-starter-webmvc-ui:2.8.5'
+	implementation 'org.springframework.boot:spring-boot-starter-web'
+	implementation 'io.swagger.core.v3:swagger-annotations:2.2.15'
+	implementation 'org.springframework.boot:spring-boot-starter-validation'
```
Вместо `'org.springframework.boot:spring-boot-starter-web'` и `'org.springframework.boot:spring-boot-starter-validation'` опять же лучше подключить только аннотации

E-Line21 commented on services/user-service/user-impl/build.gradle:
```diff
@@ -0,0 +1,27 @@
+    testImplementation 'org.mockito:mockito-junit-jupiter'
+    testImplementation("org.testcontainers:junit-jupiter")
+
+    implementation 'io.github.cdimascio:dotenv-java:3.0.0'
```
А зачем нужна эта библиотека? Я не нашел, где она используется😁

rbrmnv committed:
add postgres test-container integration tests
fix UserEntity.role processing

rbrmnv committed:
/users?email=<email> now also returns a password hash

rbrmnv committed:
change general endpoint for API

rbrmnv committed:
Merge branch 'develop' into feature/BRAEN-06/user-service-dev
